internal/faults/common.go

```diff
@@ -0,0 +1,135 @@
+// Package faults provides common headers and client-side fault injection
+// functionality.
+package faults
+
+import (
+	"fmt"
+	"log/slog"
+	"math/rand/v2"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// GetHeaderFn is the function type to return the value of a protocol-specific
+// header with the given key.
+type GetHeaderFn func(key string) string
+
+// ResumeFn is the function type to continue processing the protocol-specific
+// request without injecting a fault.
+type ResumeFn[T any] func() (T, error)
+
+// ResponseFn is the function type to inject a protocol-specific fault with the
+// given code and message.
+type ResponseFn[T any] func(code int, message string) (T, error)
+
+// SleepFn is the function type to sleep for the given duration. It is only
+// exposed for testing purposes.
+type SleepFn func(d time.Duration)
+
+func getPercentage(percentageHeader string, GetHeaderFn func(key string) string) (int, error) {
+	percentageStr := GetHeaderFn(percentageHeader)
+	if percentageStr == "" {
+		return 100, nil
+	}
+	percentage, err := strconv.Atoi(percentageStr)
+	if err != nil {
+		return 0, fmt.Errorf("provided percentage %q is not a valid integer: %w", percentageStr, err)
+	}
+	if percentage < 0 || percentage > 100 {
+		return 0, fmt.Errorf("provided percentage %q is outside the valid range of [0-100]", percentage)
+	}
+	return percentage, nil
+}
+
+type InjectFaultParams[T any] struct {
+	CallerName string
+
+	Address, Method            string
+	AbortCodeMin, AbortCodeMax int
+
+	GetHeaderFn GetHeaderFn
+	ResumeFn    ResumeFn[T]
+	ResponseFn  ResponseFn[T]
+
+	// Exposed for tests.
+	RandInt *int
+	SleepFn *SleepFn
```

Comment:
for tests you don't need to export them, just don't use external test.

Reply:
Done.